docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,410 @@
+metadata:
+name: example-1
+namespace: vsphere-site1
+spec:
+ cloneMode: FullClone
+ numCPUs: 8
+ memoryMiB: 8192
+ diskGiB: 45
+ network:
+   devices:
+   - dhcp4: false
+     fromPool: # reference to the pool
+       group: ipam.cluster.x-k8s.io/v1alpha1
+       kind: IPPool
+       name: testpool
+status:
+ network:
+   devices:
+     - claim:
+         name: example-1-1
+```
+
+### User Stories
+
+- Detail the things that people will be able to do if this proposal is implemented.
+- Include as much detail as possible so that people can understand the "how" of the system.
+- The goal here is to make this feel real for users without getting bogged down.
+
+#### Story 1
+
+As a CAPI user I want to use static IP addresses with my Machines to avoid problems with DHCP lease duration during rolling cluster upgrades.
+
+#### Story 2
+
+As a CAPI user I want to integrate machines provisioned by CAPI with an external IPAM system and allocate IP addresses from it.
+
+#### Story 3
+
+As an infrastructure provider I want a single interface to integrate with multiple different IPAM providers.
+
+#### Story 4
+
+As an IPAM provider I want a single interface to integrate with all infrastructure providers that can benefit from IPAM integration.
+
+<!--
+### Requirements (Optional)
+
+Some authors may wish to use requirements in addition to user stories.
+Technical requirements should derived from user stories, and provide a trace from
+use case to design, implementation and test case. Requirements can be prioritised
+using the MoSCoW (MUST, SHOULD, COULD, WON'T) criteria.
+
+The FR and NFR notation is intended to be used as cross-references across a CAEP.
+
+The difference between goals and requirements is that between an executive summary
+and the body of a document. Each requirement should be in support of a goal,
+but narrowly scoped in a way that is verifiable or ideally - testable.
+
+#### Functional Requirements
+
+Functional requirements are the properties that this design should include.
+
+##### FR1
+
+##### FR2
+
+#### Non-Functional Requirements
+
+Non-functional requirements are user expectations of the solution. Include
+considerations for performance, reliability and security.
+
+##### NFR1
+
+##### NFR2
+-->
+
+### Implementation Details/Notes/Constraints
+
+<!--
+- What are some important details that didn't come across above.
+- What are the caveats to the implementation?
+- Go in to as much detail as necessary here.
+- Talk about core concepts and how they releate.
+-->
+
+- Only the custom resource definitions for IPAddress and IPAddressClaim, and relevant webhooks should become part of the main cluster-api project
+- An example provider that provides in-cluster IP Address management (also useful for testing, or when no other IPAM solution is available) should be implemented as a separate project in a separate repository.
+- A shared library for providers and consumers should be implemented as part of that extra project.
```

Comment:
I will once I figure that out exactly :smile: I'm working on a reference in-cluster implementation, which might help with that. Not sure if it's necessary/helpful to try to design this beforehand, or if it's easier to build it when creating the first few implementations.

Reply:
Makes sense; If possible we should document high level the capabilities we expect this library should offer, otherwise let's just add a note that details will be added in a follow up iteration